wfe2/wfe.go

```diff
@@ -1275,3 +1279,74 @@ func (wfe *WebFrontEndImpl) addIssuingCertificateURLs(response http.ResponseWrit
 	}
 	return nil
 }
+
+type orderJSON struct {
+	Status         core.AcmeStatus
+	Expires        time.Time
+	CSR            core.JSONBuffer
+	Authorizations []string
+}
+
+// NewCertificate is used by clients to create a new order object from a CSR
+func (wfe *WebFrontEndImpl) NewOrder(ctx context.Context, logEvent *requestEvent, response http.ResponseWriter, request *http.Request) {
+	body, reg, prob := wfe.validPOSTForAccount(request, ctx, logEvent)
+	addRequesterHeader(response, logEvent.Requester)
+	if prob != nil {
+		// validPOSTForAccount handles its own setting of logEvent.Errors
+		wfe.sendError(response, logEvent, prob, nil)
+		return
+	}
+
+	var rawCSR core.RawCertificateRequest
+	err := json.Unmarshal(body, &rawCSR)
+	if err != nil {
+		logEvent.AddError("unable to JSON unmarshal order request: %s", err)
+		wfe.sendError(response, logEvent, probs.Malformed("Error unmarshaling order request"), err)
+		return
+	}
+	// Assuming a properly formatted CSR there should be two four byte SEQUENCE
+	// declarations then a two byte integer declaration which defines the version
+	// of the CSR. If those two bytes (at offset 8 and 9) and equal to 2 and 0
+	// then the CSR was generated by a pre-1.0.2 version of OpenSSL with a client
+	// which didn't explicitly set the version causing the integer to be malformed
+	// and encoding/asn1 will refuse to parse it. If this is the case exit early
+	// with a more useful error message.
+	if len(rawCSR.CSR) >= 10 && rawCSR.CSR[8] == 2 && rawCSR.CSR[9] == 0 {
+		logEvent.AddError("Pre-1.0.2 OpenSSL malformed CSR")
+		wfe.sendError(
+			response,
+			logEvent,
+			probs.Malformed("CSR generated using a pre-1.0.2 OpenSSL with a client that doesn't properly specify the CSR version. See https://community.letsencrypt.org/t/openssl-bug-information/19591"),
+			nil,
+		)
+		return
+	}
+
+	order, err := wfe.RA.NewOrder(ctx, &rapb.NewOrderRequest{
+		RegistrationID: &reg.ID,
+		Csr:            rawCSR.CSR,
+	})
+	if err != nil {
+		logEvent.AddError("unable to create order: %s", err)
+		wfe.sendError(response, logEvent, problemDetailsForError(err, "Error creating new order"), err)
+		return
+	}
+
+	respObj := orderJSON{
+		Status:         core.AcmeStatus(*order.Status),
+		Expires:        time.Unix(0, *order.Expires).Truncate(time.Second).UTC(),
+		CSR:            core.JSONBuffer(order.Csr),
+		Authorizations: make([]string, len(order.Authorizations)),
+	}
+	for i, authz := range order.Authorizations {
+		respObj.Authorizations[i] = wfe.relativeEndpoint(request, authzPath+string(*authz.Id))
+	}
+
+	response.Header().Set("Location", wfe.relativeEndpoint(request, fmt.Sprintf("%s%d", orderPath, order.Id)))
```

Comment:
Can you add a `TODO(@roland)` or a comment of some kind here about adding a `wfe.HandleFunc` for the `orderPath`? I think right now the Location header would be set but you couldn't use it.